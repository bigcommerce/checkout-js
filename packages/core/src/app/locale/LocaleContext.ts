import { CurrencyService, LanguageService } from '@bigcommerce/checkout-sdk';
import { createContext } from 'react';

export interface LocaleContextType {
    language: LanguageService;
    date?: {
        inputFormat: string;
    };
    currency?: CurrencyService;
}

const LocaleContext = createContext<LocaleContextType | undefined>(undefined);

export default LocaleContext;
