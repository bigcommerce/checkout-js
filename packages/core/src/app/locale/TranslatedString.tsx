import React, { Fragment, FunctionComponent } from 'react';

import withLanguage, { WithLanguageProps } from './withLanguage';

export interface TranslatedStringProps {
    id: string;
    data?: any;
}

const TranslatedString: FunctionComponent<TranslatedStringProps & WithLanguageProps> = ({
    data,
    id,
    language,
}) => (
    <Fragment>
        { language.translate(id, data) }
    </Fragment>
);

export default withLanguage(TranslatedString);
