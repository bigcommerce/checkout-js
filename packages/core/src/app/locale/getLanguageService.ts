import { createLanguageService, LanguageConfig, LanguageService } from '@bigcommerce/checkout-sdk';

import { FALLBACK_LOCALE, FALLBACK_TRANSLATIONS } from './translations';

let languageService: LanguageService | undefined;

export default function getLanguageService(): LanguageService {
    languageService =
        languageService ??
        createLanguageService({
            fallbackLocale: FALLBACK_LOCALE,
            fallbackTranslations: FALLBACK_TRANSLATIONS,
        });

    return languageService;
}

export type InitializeLanguageService = typeof initializeLanguageService;

export function initializeLanguageService(config: LanguageConfig): LanguageService {
    languageService = createLanguageService({
        ...config,
        defaultLocale: config.locale,
        defaultTranslations: config.defaultTranslations,
        fallbackLocale: FALLBACK_LOCALE,
        fallbackTranslations: FALLBACK_TRANSLATIONS,
    });

    return languageService;
}
