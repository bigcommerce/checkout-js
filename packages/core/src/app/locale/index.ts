export { default as LocaleContext, LocaleContextType } from './LocaleContext';
export { default as createLocaleContext } from './createLocaleContext';
export { default as getDefaultTranslations } from './getDefaultTranslations';
export { default as getLanguageService, InitializeLanguageService } from './getLanguageService';
export { default as masterpassFormatLocale } from './masterpassFormatLocale';
export { default as withCurrency, WithCurrencyProps } from './withCurrency';
export { default as withLanguage, WithLanguageProps } from './withLanguage';
export { default as withDate, WithDateProps } from './withDate';
export { default as LanguageWindow, isLanguageWindow } from './LanguageWindow';
export { default as LocaleProvider } from './LocaleProvider';
export { default as TranslatedHtml, TranslatedHtmlProps } from './TranslatedHtml';
export { default as TranslatedLink, TranslatedLinkProps } from './TranslatedLink';
export { default as TranslatedString, TranslatedStringProps } from './TranslatedString';
