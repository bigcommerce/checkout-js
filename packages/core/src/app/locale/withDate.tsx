import { createInjectHoc, InjectHoc } from '../common/hoc';

import LocaleContext from './LocaleContext';

export interface WithDateProps {
    date: {
        inputFormat: string;
    };
}

const withDate: InjectHoc<WithDateProps> = createInjectHoc(LocaleContext, {
    displayNamePrefix: 'withDate',
    pickProps: (value, key) => key === 'date' && !!value,
});

export default withDate;
