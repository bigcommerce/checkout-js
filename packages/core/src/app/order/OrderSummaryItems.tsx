import { LineItemMap } from '@bigcommerce/checkout-sdk';
import React, { Fragment, ReactNode } from 'react';

import { TranslatedString } from '../locale';
import { IconChevronDown, IconChevronUp } from '../ui/icon';

import getItemsCount from './getItemsCount';
import mapFromCustom from './mapFromCustom';
import mapFromDigital from './mapFromDigital';
import mapFromGiftCertificate from './mapFromGiftCertificate';
import mapFromPhysical from './mapFromPhysical';
import OrderSummaryItem from './OrderSummaryItem';

const COLLAPSED_ITEMS_LIMIT = 4;

export interface OrderSummaryItemsProps {
    items: LineItemMap;
}

interface OrderSummaryItemsState {
    isExpanded: boolean;
}

class OrderSummaryItems extends React.Component<OrderSummaryItemsProps, OrderSummaryItemsState> {
    constructor(props: OrderSummaryItemsProps) {
        super(props);

        this.state = {
            isExpanded: false,
        };
    }

    render(): ReactNode {
        const { items } = this.props;
        const { isExpanded } = this.state;

        return (<Fragment>
            <h3
                className="cart-section-heading optimizedCheckout-contentPrimary"
                data-test="cart-count-total"
            >
                <TranslatedString
                    data={ { count: getItemsCount(items) } }
                    id="cart.item_count_text"
                />
            </h3>

            <ul aria-live="polite" className="productList">
                {
                    [
                        ...items.physicalItems
                            .slice()
                            .sort(item => item.variantId)
                            .map(mapFromPhysical),
                        ...items.giftCertificates
                            .slice()
                            .map(mapFromGiftCertificate),
                        ...items.digitalItems
                            .slice()
                            .sort(item => item.variantId)
                            .map(mapFromDigital),
                        ...(items.customItems || [])
                            .map(mapFromCustom),
                    ]
                        .slice(0, isExpanded ? undefined : COLLAPSED_ITEMS_LIMIT)
                        .map(summaryItemProps =>
                            <li
                                className="productList-item is-visible"
                                key={ summaryItemProps.id }
                            >
                                <OrderSummaryItem { ...summaryItemProps } />
                            </li>
                        )
                }
            </ul>

            { this.renderActions() }
        </Fragment>);
    }

    private renderActions(): ReactNode {
        const { isExpanded } = this.state;

        if (this.getLineItemCount() < 5) {
            return;
        }

        return (
            <div className="cart-actions">
                <button
                    className="button button--tertiary button--tiny optimizedCheckout-buttonSecondary"
                    onClick={ this.handleToggle }
                    type="button"
                >
                        { isExpanded ?
                            <Fragment>
                                <TranslatedString id="cart.see_less_action" />
                                <IconChevronUp />
                            </Fragment> :
                            <Fragment>
                                <TranslatedString id="cart.see_all_action" />
                                <IconChevronDown />
                            </Fragment> }
                </button>
            </div>
        );
    }

    private getLineItemCount(): number {
        const { items } = this.props;

        return (items.customItems || []).length +
            items.physicalItems.length +
            items.digitalItems.length +
            items.giftCertificates.length;
    }

    private handleToggle: () => void = () => {
        const { isExpanded } = this.state;

        this.setState({ isExpanded: !isExpanded });
    };
}

export default OrderSummaryItems;
