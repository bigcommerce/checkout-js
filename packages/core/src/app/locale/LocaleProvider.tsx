import { CheckoutService, createCurrencyService, StoreConfig } from '@bigcommerce/checkout-sdk';
import { memoizeOne } from '@bigcommerce/memoize';
import React, { Component, ReactNode } from 'react';

import getLanguageService from './getLanguageService';
import LocaleContext from './LocaleContext';

export interface LocaleProviderProps {
    checkoutService: CheckoutService;
}

export interface LocaleProviderState {
    config?: StoreConfig;
}

class LocaleProvider extends Component<LocaleProviderProps> {
    state: Readonly<LocaleProviderState> = {};

    private languageService = getLanguageService();
    private unsubscribe?: () => void;

    private getContextValue = memoizeOne((config?: StoreConfig) => {
        return {
            currency: config ? createCurrencyService(config) : undefined,
            date: config
                ? {
                      inputFormat: config.inputDateFormat,
                  }
                : undefined,
            language: this.languageService,
        };
    });

    componentDidMount(): void {
        const { checkoutService } = this.props;

        this.unsubscribe = checkoutService.subscribe(
            ({ data }) => {
                this.setState({ config: data.getConfig() });
            },
            ({ data }) => data.getConfig(),
        );
    }

    componentWillUnmount(): void {
        if (this.unsubscribe) {
            this.unsubscribe();
            this.unsubscribe = undefined;
        }
    }

    render(): ReactNode {
        const { children } = this.props;
        const { config } = this.state;

        return (
            <LocaleContext.Provider value={this.getContextValue(config)}>
                {children}
            </LocaleContext.Provider>
        );
    }
}

export default LocaleProvider;
