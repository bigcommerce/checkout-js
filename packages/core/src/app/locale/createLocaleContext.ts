import { createCurrencyService, StoreConfig } from '@bigcommerce/checkout-sdk';

import getLanguageService from './getLanguageService';
import { LocaleContextType } from './LocaleContext';

export default function createLocaleContext(config: StoreConfig): Required<LocaleContextType> {
    if (!config) {
        throw new Error('Missing configuration data');
    }

    const { inputDateFormat } = config;

    return {
        currency: createCurrencyService(config),
        date: {
            inputFormat: inputDateFormat,
        },
        language: getLanguageService(),
    };
}
