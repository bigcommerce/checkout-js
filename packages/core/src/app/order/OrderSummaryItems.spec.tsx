import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import React from 'react';

import {
    getCustomItem,
    getDigitalItem,
    getGiftCertificateItem,
    getPhysicalItem,
} from '../cart/lineItem.mock';
import { getStoreConfig } from '../config/config.mock';
import { createLocaleContext, LocaleContext, LocaleContextType, TranslatedString } from '../locale';

import OrderSummaryItems from './OrderSummaryItems';

describe('OrderSummaryItems', () => {
    describe('when it has 4 line items or less', () => {
        let orderSummaryItems: ShallowWrapper;

        beforeEach(() => {
            orderSummaryItems = shallow(
                <OrderSummaryItems
                    items={{
                        customItems: [getCustomItem()],
                        physicalItems: [getPhysicalItem()],
                        digitalItems: [getDigitalItem()],
                        giftCertificates: [getGiftCertificateItem()],
                    }}
                />,
            );
        });

        it('renders total count', () => {
            expect(orderSummaryItems.find(TranslatedString).props()).toMatchObject({
                id: 'cart.item_count_text',
                data: { count: 5 },
            });
        });

        it('renders product list', () => {
            expect(orderSummaryItems.find('.productList')).toMatchSnapshot();
        });

        it('does not render actions', () => {
            expect(orderSummaryItems.find('.cart-actions')).toHaveLength(0);
        });
    });

    describe('when it has 5 line items or more', () => {
        let localeContext: LocaleContextType;
        let orderSummaryItems: ReactWrapper;

        beforeEach(() => {
            localeContext = createLocaleContext(getStoreConfig());

            orderSummaryItems = mount(
                <LocaleContext.Provider value={localeContext}>
                    <OrderSummaryItems
                        items={{
                            customItems: [getCustomItem()],
                            physicalItems: [
                                {
                                    ...getPhysicalItem(),
                                    id: '664',
                                },
                                getPhysicalItem(),
                            ],
                            digitalItems: [getDigitalItem()],
                            giftCertificates: [getGiftCertificateItem()],
                        }}
                    />
                </LocaleContext.Provider>,
            );
        });

        it('renders actions', () => {
            expect(orderSummaryItems.find('.cart-actions')).toHaveLength(1);
        });

        it('is collapsed by default', () => {
            expect(orderSummaryItems.find('.cart-actions').text()).toMatch('See All');

            expect(orderSummaryItems.find('.productList-item')).toHaveLength(4);
        });

        describe('when action is clicked', () => {
            beforeEach(() => {
                orderSummaryItems.find('.cart-actions button').simulate('click');
            });

            it('shows the rest of the items', () => {
                expect(orderSummaryItems.find('.cart-actions').text()).toMatch('See Less');

                expect(orderSummaryItems.find('.productList-item')).toHaveLength(5);
            });

            describe('when action is clicked a second time', () => {
                beforeEach(() => {
                    orderSummaryItems.find('.cart-actions button').simulate('click');
                });

                it('collapses line items back', () => {
                    expect(orderSummaryItems.find('.cart-actions').text()).toMatch('See All');

                    expect(orderSummaryItems.find('.productList-item')).toHaveLength(4);
                });
            });
        });
    });
});
