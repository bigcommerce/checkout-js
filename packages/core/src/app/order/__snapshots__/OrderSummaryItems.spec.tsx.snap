// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderSummaryItems when it has 4 line items or less renders product list 1`] = `
<ul
  aria-live="polite"
  className="productList"
>
  <li
    className="productList-item is-visible"
    key="666"
  >
    <Memo(OrderSummaryItem)
      amount={250}
      amountAfterDiscount={200}
      id="666"
      image={
        <img
          alt="Canvas Laundry Cart"
          data-test="cart-item-image"
          src="/images/canvas-laundry-cart.jpg"
        />
      }
      name="Canvas Laundry Cart"
      productOptions={
        Array [
          Object {
            "content": "n v",
            "testId": "cart-item-product-option",
          },
        ]
      }
      quantity={1}
    />
  </li>
  <li
    className="productList-item is-visible"
    key="bd391ead-8c58-4105-b00e-d75d233b429a"
  >
    <Memo(OrderSummaryItem)
      amount={100}
      id="bd391ead-8c58-4105-b00e-d75d233b429a"
      image={
        <span
          className="productImage-giftCertificate"
          data-test="cart-item-gift-certificate"
        >
          <IconGiftCertificate />
        </span>
      }
      name="$100 Gift Certificate"
      quantity={1}
    />
  </li>
  <li
    className="productList-item is-visible"
    key="667"
  >
    <Memo(OrderSummaryItem)
      amount={200}
      amountAfterDiscount={200}
      id="667"
      image={
        <img
          alt="Digital Book"
          data-test="cart-item-image"
          src="/images/digital-book.jpg"
        />
      }
      name="Digital Book"
      productOptions={
        Array [
          Object {
            "content": "m l",
            "testId": "cart-item-product-option",
          },
          Object {
            "content": <a
              href="url.php"
              rel="noopener noreferrer"
              target="_blank"
            >
              <WithLanguage(TranslatedString)
                id="cart.downloads_action"
              />
            </a>,
            "testId": "cart-item-digital-product-download",
          },
        ]
      }
      quantity={1}
    />
  </li>
  <li
    className="productList-item is-visible"
    key="55e11c8f-7dce-4da3-9413-b649533f8bad"
  >
    <Memo(OrderSummaryItem)
      amount={20}
      id="55e11c8f-7dce-4da3-9413-b649533f8bad"
      name="Custom item"
      quantity={2}
    />
  </li>
</ul>
`;
